pkg/kapp/resourcesmisc/custom_waiting_resource.go

```diff
@@ -81,16 +85,41 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
 	hasConditionWaitingForGeneration := false
 	// Check on failure conditions first
 	for _, condMatcher := range s.waitRule.ConditionMatchers {
-		for _, cond := range obj.Status.Conditions {
-			if cond.Type == condMatcher.Type && cond.Status == condMatcher.Status {
+		// Check whether timeout has occured
+		var isTimeOutConditionPresent bool
+		if len(condMatcher.Timeout) != 0 {
+			for _, cond := range obj.Status.Conditions {
 				if condMatcher.SupportsObservedGeneration && obj.Metadata.Generation != cond.ObservedGeneration {
 					hasConditionWaitingForGeneration = true
 					continue
 				}
-				if condMatcher.Failure {
-					return DoneApplyState{Done: true, Successful: false, Message: fmt.Sprintf(
-						"Encountered failure condition %s == %s: %s (message: %s)",
-						cond.Type, condMatcher.Status, cond.Reason, cond.Message)}
+				if condMatcher.Type == cond.Type && cond.Status == condMatcher.Status {
+					isTimeOutConditionPresent = true
+					if s.hasTimeoutOccurred(condMatcher.Timeout, fmt.Sprintf("%s.%s", s.resource.Namespace(), s.resource.Name())) {
+						return DoneApplyState{Done: true, Successful: false, Message: fmt.Sprintf(
+							"Encountered failure condition %s == %s: %s (message: %s) continuously for %s duration",
+							cond.Type, condMatcher.Status, cond.Reason, cond.Message, condMatcher.Timeout)}
+					}
+					return DoneApplyState{Done: false, Message: fmt.Sprintf(
+						"%s: %s (message: %s)",
+						cond.Type, cond.Reason, cond.Message)}
+				}
+			}
+			if !isTimeOutConditionPresent {
+				timeoutMap.Delete(fmt.Sprintf("%s.%s", s.resource.Namespace(), s.resource.Name()))
+			}
+		} else {
```

Comment:
nit: Let's return early and drop the else block?

Reply:
Adopted